contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+# Best Practices with `git`
+
+Author(s): [Darren Wee](https://github.com/darrenwee)
+
+- [Introduction](#introduction)
+- [Best Practices](#best-practices)
+    - [Write Good Commit Messages](#write-good-commit-messages)
+        - [What Constitutes a Good Commit Message](#what-constitutes-a-good-commit-message)
+        - [Set Up Your Editor for Commit Messages](#set-up-your-editor-for-commit-messages)
+    - [Always Commit Functional Code](#always-commit-functional-code)
+        - [Stashing](#stashing)
+    - [One Logical Change to One Commit](#one-logical-change-to-one-commit)
+    - [Hide the Sausage Making](#hide-the-sausage-making)
+        - [How to Hide the Sausage](#how-to-hide-the-sausage)
+    - [Respect Published History](#respect-published-history)
+        - [Exceptions](#exceptions)
+    - [Keep Up To Date](#keep-up-to-date)
+        - [Use Remotes Effectively](#use-remotes-effectively)
+        - [Rebase versus Merging](#rebase-versus-merging)
+- [Resources](#resources)
+    - [Further Reading](#further-reading)
+
+
+# Introduction
+[`git`](https://git-scm.com/) is a popular source code management tool and commonly used in many open-source projects, especially those on [GitHub](https://github.com).
+
+# Best Practices
+`git` can be an incredibly useful tool for collaboration or it can be a terrible headache. Best practices exist in order to create a common understanding between users so that the latter does not happen.
+
+Best practices are guidelines that are mostly sensible, but are still guidelines. You can always choose to ignore them although best if you have a compelling reason to do so.
+
+---
+
+## Write Good Commit Messages
+Good commit messages can help:
+- reviewers to understand the high-level changes in your pull request/patch
+- others to understand your reasoning behind changes made by that commit,
+    - while reviewing your code,
+    - while figuring out why a piece of code that is five years old is that way,
+
+### What Constitutes a Good Commit Message
+The easiest way to attain commit message discipline is to stop putting in one-liner descriptions using `git commit -m "Add some things to that."`. Instead, use `git commit` and write a proper message in an editor.
+
+A good commit message can be formatted the following way:
+```
+Capitalized, short (50 chars or less) summary
+
+More detailed explanatory text, if necessary.  Wrap it to about 72
+characters or so.  In some contexts, the first line is treated as the
+subject of an email and the rest of the text as the body.  The blank
+line separating the summary from the body is critical (unless you omit
+the body entirely); tools like rebase can get confused if you run the
+two together.
+
+Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
```

Comment:
One good thing I have read in many articles related to writing git messages is to make sure the message makes sense when read as follows:

>If applied, this commit will `your commit message here`

Maybe can add this here?

Reply:
I added it outside of the commit message code block so that it stands out more.